File: android/src/main/java/com/admanager/CustomNativeAdLoader.java
package com.admanager;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.UiThreadUtil;
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdLoader;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.VideoOptions;
import com.google.android.gms.ads.admanager.AdManagerAdRequest;
import com.google.android.gms.ads.nativead.NativeAdOptions;
import com.google.android.gms.ads.nativead.NativeCustomFormatAd;

import java.util.UUID;

public class CustomNativeAdLoader {

  private String loaderId;
  private String adUnitId;
  private String formatId;
  private CustomNativeAdState _adState;
  private CustomNativeAdError error;

  private NativeAdOptions.Builder adOptionBuilder;
  private AdLoader.Builder adLoaderBuilder;
  private AdLoader adLoader;
  private AdManagerAdRequest adRequest;
  private NativeCustomFormatAd receivedAd;

  private NativeCustomFormatAd.OnCustomClickListener customClickListener;

  private ReactApplicationContext reactApplicationContext;

  private AdStateChangeListener adStateChangeListener;
  private CustomNativeAdLoaderHandler adLoaderCompletionListener;

  public interface AdStateChangeListener {
    void onStateChanged(CustomNativeAdLoader loader, CustomNativeAdState oldState, CustomNativeAdState newState);
  }

  CustomNativeAdLoader(ReactApplicationContext context, String adUnitId, String formatId) {
    UUID uuid = UUID.randomUUID();
    this.loaderId = uuid.toString();

    this.adUnitId = adUnitId;
    this.formatId = formatId;
    this.reactApplicationContext = context;
    this.error = null;
    this._adState = CustomNativeAdState.CustomNativeAdStateInit;
    this.initBuilder();
  }

  private void initBuilder() {
    this.adLoaderBuilder = new AdLoader.Builder(this.reactApplicationContext, adUnitId);
    this.adOptionBuilder = new NativeAdOptions.Builder();
  }

  private void forceUpdateState(CustomNativeAdState newState) {
    CustomNativeAdState oldState = this._adState;
    this._adState = newState;
    if (this.adStateChangeListener != null) {
      this.adStateChangeListener.onStateChanged(this, oldState, newState);
    }
  }

  private void updateState(CustomNativeAdState newState) {
    CustomNativeAdState oldState = this._adState;
    if(oldState == CustomNativeAdState.CustomNativeAdStateError || oldState.getValue() < newState.getValue()) {
      this.forceUpdateState(newState);
    }
  }

  public String getAdUnitId() {
    return this.adUnitId;
  }

  public String getLoaderId() {
    return this.loaderId;
  }

  public String getFormatId() {
    return this.formatId;
  }

  public Boolean hasCustomClickHandler() {
    return this.customClickListener != null;
  }

  public void setVideoOptions(@Nullable VideoOptions options) {
    // Methods in the NativeAdOptions.Builder class can be
    // used here to specify individual options settings.
    if (options == null) {
      adOptionBuilder.setVideoOptions(new VideoOptions.Builder().setStartMuted(true).build());
    } else {
      adOptionBuilder.setVideoOptions(options);
    }
  }
  public void setReturnUrlsForImageAssets(Boolean shouldSet) {
    this.adOptionBuilder.setReturnUrlsForImageAssets(shouldSet);
  }

  public void removeCustomClickHandler() throws CustomNativeAdError {
    if (this._adState.getValue() <= CustomNativeAdState.CustomNativeAdStateLoading.getValue()) {
      this.customClickListener = null;
    } else {
      throw CustomNativeAdError.withMessage("Custom click handler can only be set before the loader starts to fetch an ad.", "SET_CUSTOM_CLICK_AFTER_LOAD");
    }
  }
  public void setCustomClickHandler(NativeCustomFormatAd.OnCustomClickListener handler) throws CustomNativeAdError {
    if (this._adState.getValue() <= CustomNativeAdState.CustomNativeAdStateLoading.getValue()) {
      this.customClickListener = handler;
    } else {
      throw CustomNativeAdError.withMessage("Custom click handler can only be set before the loader starts to fetch an ad.", "SET_CUSTOM_CLICK_AFTER_LOAD");
    }
  }

  public void clearup() {
    this.customClickListener = null;
    if (this.receivedAd != null) {
      this.receivedAd.destroy();
      this.receivedAd = null;
    }
    this.adLoader = null;
    this.adRequest = null;
    this.initBuilder();
    this.setVideoOptions(null);
    this.forceUpdateState(CustomNativeAdState.CustomNativeAdStateInit);
  }



  public CustomNativeAdLoaderDetails getDetails() {
    CustomNativeAdLoaderDetails details = new CustomNativeAdLoaderDetails(this.loaderId, this.adUnitId, this.formatId, this._adState);
    details.setReceivedAd(this.receivedAd);
    return details;
  }

  private void prepareForAdLoading() {
    this.adLoaderBuilder.forCustomFormatAd(this.getFormatId(), new NativeCustomFormatAd.OnCustomFormatAdLoadedListener() {
      @Override
      public void onCustomFormatAdLoaded(@NonNull NativeCustomFormatAd nativeCustomFormatAd) {
        // Show the custom format and record an impression.
        CustomNativeAdLoader.this.receivedAd = nativeCustomFormatAd;
        CustomNativeAdLoader.this.updateState(CustomNativeAdState.CustomNativeAdStateReceived);
        CustomNativeAdLoader.this.adLoaderCompletionListener.onAdReceived(CustomNativeAdLoader.this, nativeCustomFormatAd);
        CustomNativeAdLoader.this.adLoaderCompletionListener = null;
        // If this callback occurs after the activity is destroyed, you
        // must call destroy and return or you may get a memory leak.
        // Note `isDestroyed()` is a method on Activity.
        if (CustomNativeAdLoader.this.reactApplicationContext.getCurrentActivity().isDestroyed()) {
          nativeCustomFormatAd.destroy();
          return;
        }
      }
    }, this.customClickListener);
    this.adLoaderBuilder.withNativeAdOptions(CustomNativeAdLoader.this.adOptionBuilder.build());
    this.adLoaderBuilder.withAdListener(new AdListener() {
      @Override
      public void onAdFailedToLoad(@NonNull LoadAdError var1) {
        CustomNativeAdError error = CustomNativeAdError.fromAdError(var1, "FAILED_TO_RECEIVE_AD");
        CustomNativeAdLoader.this.error = error;
        CustomNativeAdLoader.this.updateState(CustomNativeAdState.CustomNativeAdStateError);
        // Handle the failure by logging, altering the UI, and so on.
        CustomNativeAdLoader.this.adLoaderCompletionListener.onAdLoadFailed(CustomNativeAdLoader.this, error);
        CustomNativeAdLoader.this.adLoaderCompletionListener = null;
      }
    });
  }

  public void loadAd(AdManagerAdRequest adRequest, CustomNativeAdLoaderHandler completionListener) {
    UiThreadUtil.runOnUiThread(new Runnable() {
      @Override
      public void run() {
        if(adLoader != null && adLoader.isLoading()){
          CustomNativeAdError error = CustomNativeAdError.withMessage("This loader is already loading a request.", "AD_REQUEST_ALREADY_RUNNING");
          CustomNativeAdLoader.this.error = error;
          CustomNativeAdLoader.this.updateState(CustomNativeAdState.CustomNativeAdStateError);
          CustomNativeAdLoader.this.adLoaderCompletionListener.onAdLoadFailed(CustomNativeAdLoader.this, error);
        }else{
          if (adLoader == null) {
            CustomNativeAdLoader.this.prepareForAdLoading();
          }
          CustomNativeAdLoader.this.adLoaderCompletionListener = completionListener;
          CustomNativeAdLoader.this.adLoader = adLoaderBuilder.build();
          CustomNativeAdLoader.this.adRequest = adRequest;
          CustomNativeAdLoader.this.updateState(CustomNativeAdState.CustomNativeAdStateLoading);
          CustomNativeAdLoader.this.adLoader.loadAd(adRequest);
        }
      }
    });
  }

  public void displayAd() throws CustomNativeAdError {
    if (this._adState.getValue() >= CustomNativeAdState.CustomNativeAdStateReceived.getValue()) {
      this.updateState(CustomNativeAdState.CustomNativeAdStateDisplaying);
    } else {
      throw CustomNativeAdError.withMessage("The ad is not ready to display.", "AD_NOT_DISPLAYABLE");
    }
  }

  public void recordImpression() throws CustomNativeAdError {
    if (this.receivedAd != null) {
      UiThreadUtil.runOnUiThread(new Runnable() {
        @Override
        public void run() {
          CustomNativeAdLoader.this.receivedAd.recordImpression();
        }
      });
      CustomNativeAdLoader.this.updateState(CustomNativeAdState.CustomNativeAdStateImpression);
    }else{
      throw CustomNativeAdError.withMessage("Could not record impression. Propably no ad received yet.", "RECORD_AD_IMPRESSION_FAILED");
    }
  }

  public String recordClick(@Nullable String assetKey) throws CustomNativeAdError {
    if (this.receivedAd != null) {
      String clickedAssetKey;
      if(assetKey == null){
        clickedAssetKey = Utils.getOneAssetKey(this.receivedAd);
      }else{
        clickedAssetKey = assetKey;
      }
      UiThreadUtil.runOnUiThread(new Runnable() {
        @Override
        public void run() {
          //It does not matter if we call this function to often...google will take care of that and still tracks it as a single impression
          //so if someone clicks it the ad must have been seen
          CustomNativeAdLoader.this.receivedAd.recordImpression();
          CustomNativeAdLoader.this.receivedAd.performClick(clickedAssetKey);
        }
      });
      CustomNativeAdLoader.this.updateState(CustomNativeAdState.CustomNativeAdStateClicked);
      return clickedAssetKey;
    }else{
      throw CustomNativeAdError.withMessage("Could not record click on asset key. Propably no ad received yet.", "RECORD_AD_CLICK_FAILED");
    }
  }

  public void makeOutdated() {
    this.updateState(CustomNativeAdState.CustomNativeAdStateOutdated);
  }
}

File: ios/CustomNativeAdLoader.m
//
//  CustomNativeAdLoader.m
//  react-native-ad-manager
//
//  Created by Christian Schaffrath on 18.05.23.
//

#import <GoogleMobileAds/GoogleMobileAds.h>
#import "CustomNativeAdLoader.h"
#import "GADCustomNativeAd+Infos.h"
#import <React/RCTLog.h>

@interface CustomNativeAdLoader()<GADAdLoaderDelegate, GADCustomNativeAdLoaderDelegate, GADCustomNativeAdDelegate>

@property(nonatomic) NSString * loaderId;
@property(nonatomic) NSString * adUnitId;
@property(nonatomic) NSArray<NSString *>* formatIds;
@property(nonatomic) CustomNativeAdState currentState;

@property(nonatomic) BOOL isRequestingAd;
@property(nonatomic) AdLoaderCompletionBlock completionBlock;
@property(nullable, nonatomic) GAMRequest * request;

@property (nonatomic, strong) GADAdLoader * customAdLoader;
@property (nullable, nonatomic, strong) GADCustomNativeAd *receivedAd;
@property (nullable, nonatomic, strong) CustomNativeAdError *error;

@end

@implementation CustomNativeAdLoader
{
    GADNativeAdCustomClickHandler customClickHandler;
}

+ (NSString *) generateId {
    // Generate a random UUID
    CFUUIDRef uuid = CFUUIDCreate(kCFAllocatorDefault);
    NSString *uuidString = (__bridge_transfer NSString *)CFUUIDCreateString(kCFAllocatorDefault, uuid);
    CFRelease(uuid);
    return  uuidString;
}

- (nonnull instancetype)initWithAdUnitId:(nonnull NSString *)adUnitId
                            andFormatIds: (NSArray<NSString*>*) formatIds
                      rootViewController:(nullable UIViewController *)rootViewController
                                 options:(nullable NSArray<GADAdLoaderOptions *> *)options {
    self = [super init];
    if(self)
    {
        self.loaderId = [CustomNativeAdLoader generateId];
        self.currentState = CustomNativeAdStateInit;
        self.error = nil;
        self.request = nil;
        
        self.adUnitId = adUnitId;
        self.formatIds = formatIds;
        
        GADAdLoader * googleAdLoader = [[GADAdLoader alloc] initWithAdUnitID:adUnitId
                                               rootViewController:nil //[UIApplication sharedApplication].delegate.window.rootViewController
                                                          adTypes: @[GADAdLoaderAdTypeCustomNative]
                                                          options: options];
        self.customAdLoader = googleAdLoader;
        self.customAdLoader.delegate = self;
    }
    return self;
}

- (void) updateState: (CustomNativeAdState) newState {
    CustomNativeAdState oldState = self.currentState;
    if(oldState == CustomNativeAdStateError || oldState < newState){
        [self forceUpdateState:newState];
    }
}

- (void) forceUpdateState: (CustomNativeAdState) newState {
    CustomNativeAdState oldState = self.currentState;
    self.currentState = newState;
    if(self.stateChangedDelegate){
        [self.stateChangedDelegate didChangeStateFor:self fromState:oldState toNewState:newState];
    }
}

- (void) setCustomClickHandler: (nullable GADNativeAdCustomClickHandler) handler {
    customClickHandler = handler;
    if(self.receivedAd){
        if(handler == nil){
            [self.receivedAd setCustomClickHandler:nil];
        }else{
            [self.receivedAd setCustomClickHandler:customClickHandler];
        }
    }
}

- (BOOL) hasActiveCustomClickHandler {
    return customClickHandler != nil;
}

- (NSString*) getLoaderId {
    return self.loaderId;
}

- ( GADCustomNativeAd* _Nullable ) getReceivedAd {
    return self.receivedAd;
}

- (CustomNativeAdLoaderDetails) getDetails {
    CustomNativeAdLoaderDetails details;
    
    details.loaderId = self.loaderId;
    details.adUnitId = self.adUnitId;
    details.formatIds = self.formatIds;
    details.state = self.currentState;
    
    if (self.receivedAd){
        details.hasAd = YES;
        details.ad = [self.receivedAd getAdDetails];
    }else{
        details.hasAd = NO;
    }
    return details;
}



- (void) cleanup {
    self.error = nil;
    self.completionBlock = nil;
    if(self.receivedAd){
        self.receivedAd.delegate = nil;
    }
    self.receivedAd = nil;
    [self forceUpdateState:CustomNativeAdStateInit];
}


- (void)loadRequest:(GAMRequest *)request WithCompletion:(AdLoaderCompletionBlock)completion {
    if(self.customAdLoader.loading == YES || self.isRequestingAd){
        CustomNativeAdError * error = [CustomNativeAdError errorWithMessage:@"This loader is already loading a request." andCode:@"AD_REQUEST_ALREADY_RUNNING"];
        completion(error, nil);
    }else{
        self.error = nil;
        self.request = request;
        [self updateState:CustomNativeAdStateLoading];
        self.isRequestingAd = YES;
        self.completionBlock = completion;
        self.customAdLoader.delegate = self;
        [self.customAdLoader loadRequest:self.request];
    }
}

- (BOOL) setIsDisplayingOnView: (UIView*) view {
    if (self.receivedAd){
        // Set the top-level native ad view on the GADNativeCustomTemplateAd so the
        // Google Mobile Ads SDK can track viewability for that view.
        self.receivedAd.displayAdMeasurement.view = view;
        // Begin measuring your impressions and clicks.
        NSError *error = nil;
        [self.receivedAd.displayAdMeasurement startWithError:&error];
        if (error) {
            RCTLogWarn(@"Failed to start the display measurement.For loader %@", [self getLoaderId]);
        }else{
        }
        return [self setIsDisplaying];
    }
    return NO;
}


- (BOOL) setIsDisplaying {
    if(self.receivedAd && self.currentState >= CustomNativeAdStateReceived){
        if(self.currentState <= CustomNativeAdStateDisplaying){
            [self updateState: CustomNativeAdStateDisplaying];
        }
        return YES;
    }
    return NO;
}


- (BOOL) recordImpression {
    if (self.receivedAd) {
        [self.receivedAd recordImpression];
        self.error = nil;
        [self updateState:CustomNativeAdStateImpression];
        return YES;
    }
    return NO;
}

- (BOOL) performClickOnAssetWithKey: (nonnull NSString*) assetKey {
    if (self.receivedAd){
        if (customClickHandler != nil){
            [self.receivedAd setCustomClickHandler:customClickHandler];
        }else{
            [self.receivedAd setCustomClickHandler:nil];
        }
        //It does not matter if we call this function to ofthen...google will take care of that and still tracks it as a single impression
        //so if someone clicks it the ad must have been seen
        [self.receivedAd recordImpression];
        [self.receivedAd performClickOnAssetWithKey:assetKey];
        self.error = nil;
        [self updateState:CustomNativeAdStateClicked];
        return YES;
    }
    return NO;
}

- (void) makeOutdated {
    [self updateState:CustomNativeAdStateOutdated];
}

- (void) destroy {
    [self cleanup];
}

#pragma mark GADAdLoaderDelegate implementation

/// Called after adLoader has finished loading.
- (void)adLoaderDidFinishLoading:(nonnull GADAdLoader *)adLoader {
    self.isRequestingAd = NO;
    self.request = nil;
}

- (void)adLoader:(nonnull GADAdLoader *)adLoader didFailToReceiveAdWithError:(nonnull NSError *)error {
    CustomNativeAdError *err = [CustomNativeAdError fromError:error WithCode:@"FAILED_TO_RECEIVE_AD"];
    self.error = err;
    [self updateState:CustomNativeAdStateError];
    if(self.completionBlock){
        self.completionBlock(err, nil);
        self.completionBlock = nil;
    }
    self.request = nil;
    self.isRequestingAd = NO;
}

#pragma mark GADCustomNativeAdLoaderDelegate implementation
/// Called when requesting an ad. Asks the delegate for an array of custom native ad format ID
/// strings.
- (nonnull NSArray<NSString *> *)customNativeAdFormatIDsForAdLoader:(nonnull GADAdLoader *)adLoader {
    
    if(self.formatIds != nil && [self.formatIds count] > 0) {
        return self.formatIds;
    }else{
        return @[];
    }
}

/// Tells the delegate that a custom native ad was received.
- (void)adLoader:(GADAdLoader *)adLoader didReceiveCustomNativeAd:(GADCustomNativeAd *) customNativeAd {
    self.receivedAd = customNativeAd;
    if (customClickHandler != nil){
        [self.receivedAd setCustomClickHandler:customClickHandler];
    }else{
        [self.receivedAd setCustomClickHandler:nil];
    }
    self.receivedAd.delegate = self;
    self.error = nil;
    [self updateState:CustomNativeAdStateReceived];
    if(self.completionBlock){
        self.completionBlock(nil, customNativeAd);
        self.completionBlock = nil;
    }
    self.request = nil;
    self.isRequestingAd = NO;
}

#pragma mark GADCustomNativeAdDelegate
/// Called when an impression is recorded for a custom native ad.
- (void)customNativeAdDidRecordImpression:(nonnull GADCustomNativeAd *)nativeAd {
    if(self.adActionsDelegate){
        [self.adActionsDelegate didRecordImpressionOnAd:nativeAd forAdLoader:self];
    }
}

/// Called when a click is recorded for a custom native ad.
- (void)customNativeAdDidRecordClick:(nonnull GADCustomNativeAd *)nativeAd {
    if(self.adActionsDelegate){
        [self.adActionsDelegate didRecordClickOnAd:nativeAd forAdLoader:self];
    }
}

#pragma mark Click-Time Lifecycle Notifications

/// Called just before presenting the user a full screen view, such as a browser, in response to
/// clicking on an ad. Use this opportunity to stop animations, time sensitive interactions, etc.
///
/// Normally the user looks at the ad, dismisses it, and control returns to your application with
/// the customNativeAdDidDismissScreen: message. However, if the user hits the Home button or clicks
/// on an App Store link, your application will end. The next method called will be the
/// applicationWillResignActive: of your UIApplicationDelegate object.
- (void)customNativeAdWillPresentScreen:(nonnull GADCustomNativeAd *)nativeAd {
    if(self.adActionsDelegate && [self.adActionsDelegate respondsToSelector:@selector(willPresentScreen:forAdLoader:)]){
        [self.adActionsDelegate willPresentScreen:nativeAd forAdLoader:self];
    }
}

/// Called just before dismissing a full screen view.
- (void)customNativeAdWillDismissScreen:(nonnull GADCustomNativeAd *)nativeAd{
    if(self.adActionsDelegate && [self.adActionsDelegate respondsToSelector:@selector(willDismissScreen:forAdLoader:)]){
        [self.adActionsDelegate willDismissScreen:nativeAd forAdLoader:self];
    }
}

/// Called just after dismissing a full screen view. Use this opportunity to restart anything you
/// may have stopped as part of customNativeAdWillPresentScreen:.
- (void)customNativeAdDidDismissScreen:(nonnull GADCustomNativeAd *)nativeAd {
    if(self.adActionsDelegate && [self.adActionsDelegate respondsToSelector:@selector(didDismissScreen:forAdLoader:)]){
        [self.adActionsDelegate didDismissScreen:nativeAd forAdLoader:self];
    }
}

@end
